pkg/ccl/testccl/workload/schemachange/schemachange_test: TestWorkload failed [duplicate key value violates unique constraint]

pkg/ccl/testccl/workload/schemachange/schemachange_test.TestWorkload [failed](https://tanzanite.cluster.engflow.com/invocations/default/ffe8bee2-0fbd-4151-96ac-be48464c062b?testReportRun=7&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cvY2NsL3Rlc3RjY2wvd29ya2xvYWQvc2NoZW1hY2hhbmdlOnNjaGVtYWNoYW5nZV90ZXN0) on release-25.1 @ [68bf8c8346a360a41ba90f986fb1e4c2778e048b](https://github.com/cockroachdb/cockroach/commits/68bf8c8346a360a41ba90f986fb1e4c2778e048b):


```
    "result": true
   },
   {
    "query": "SELECT ARRAY[parent.table_schema, parent.table_name, parent.column_name, child.column_name] FROM (SELECT conname, conkey, confkey, conrelid, confrelid FROM pg_constraint WHERE (contype = 'f') AND (conrelid = 'public.table_w0_68'::REGCLASS::INT8)) AS con JOIN (SELECT column_name, ordinal_position, column_default FROM information_schema.columns WHERE (table_schema = 'public') AND (table_name = 'table_w0_68')) AS child ON conkey[1] = child.ordinal_position JOIN (SELECT pc.oid, cols.table_schema, cols.table_name, cols.column_name, cols.ordinal_position FROM pg_class AS pc JOIN pg_namespace AS pn ON pc.relnamespace = pn.oid JOIN information_schema.columns AS cols ON ((pc.relname = cols.table_name) AND (pn.nspname = cols.table_schema))) AS parent ON ((con.confkey[1] = parent.ordinal_position) AND (con.confrelid = parent.oid)) WHERE child.column_name != 'rowid'",
    "result": null
   }
  ],
  "previousStatements": [
   "INSERT INTO public.table_w0_68 (col54_w0_55, col54_w0_66, col54_w0_61, col_w0_68_w1_57, col54_w0_56, col54_w0_67, col_w0_68_w1_55) VALUES ('0101000020E6100000000000000000F87F000000000000F87F':::GEOGRAPHY,NULL,'1979-09-26':::DATE,(-0.5600246728876915):::FLOAT8,e'\\x1f'::\"char\",6.751827805675903052E+27:::DECIMAL,(-5):::INT8),('0107000060E6100000070000000103000060E61000000100000009000000952F9D26000564C00C008C0249E034C0DCC03DB9A87301C252423AAC309962C062A271F3889053C0E889D1F7598ED741B1912C4B8B9E55C05EB5B7BDCC034CC0C54942EDCCDC01C2A65BE967323D6340DD565D50142054C080DBD75E7C5CA2C1CE20AC049E0A6240503CAD87397E40403DAF73CE0C96F4C194DCD482273A6440ECC8140D4C184B4099A1C8B805EB01C2F6B5089D3C7A56C080948F8DC12A0040A97D1911CDDAF6C126AD9AD8D39F65C0B87D377E681724C00C1881503517F7C1952F9D26000564C00C008C0249E034C0DCC03DB9A87301C20103000060E6100000010000000A0000007CB3D639972A54C0AC58FF0AB8F832C0685B33AC98B0EBC140B9E14E58FD64C0B4E9D6CAB9924EC056961463225DF741902825FBDE4A4DC029AF1C897CF554C0AC456D84C61C01423464C97A1E4047C0C4ACC46F6C4A38C046E675724AD8F241CC7364E22C564D406CF15C81AD8E52C054A63ABFAB24EA41A0390FB59A5B19C048D2383F0B1A22C09E8059CF7A8CEDC1881659E70F5056402C17D977E9254B40D287F4A8662BE1C1A09D6951258A11C07475ED02EBF04040D0305B203039F841DCB6D759A95066C06A8E5EEBD9FC5440080069A62292DE417CB3D639972A54C0AC58FF0AB8F832C0685B33AC98B0EBC10101000060E6100000180AD34EE5DD4FC0CAD1A0C4297A5240E866515D19B8FD410102000060E6100000020000009617923F572657C0B831223D7FCA2D40846997ACDB3EE1415853EE95F11561C07050713900CE1F40C00344FE1FDCFFC10102000060E61000000600000004D133A006CC4BC04811049E257C2640B00CFA5D71D4E74105442DE22B1E58C0D880A38B1CD424C05811816BAF27EE41AE3272173CBA61401A1078BCD1E64FC0F20DD27C324701420878A36391446240D165EE01BA5447C0C8601983F5D3D0416E4654D155E36040C02905F92C931BC094A12885325BD9C15E54CEA5083A6340BE11847ABF7C5240F83D1CD15171FF410101000060E61000005CCD7277AEE46040BA3DCE3F9F905540C0B1ACAB7982FB410105000060E61000000300000001020000400200000084922D8818CE56409F1071728A9353C02CE86214D079014234549B1333194F40CE4C89427673514090846777D51CC74101020000400200000050AC4EDE0FA631C0C8819A35E7C62540BCB6795937EDE14174F52E5304B44EC03CD2AD8B6CEF40408267F9E96464FDC101020000400500000044C7F8CC34C44F403C283EC982204FC00C32478259BBDDC130BC9496BB4E4E40D043287718231EC0F06216CAC908ED410607C0DFD17063406853712A69582DC07636C69F818CF941420085327C166340602A541FA8E34340629966226A55F7C1A0C9E30B73DD13C04C9F4A399B7550402CBDC5D8D26CE341':::GEOGRAPHY,e'/l\\x13\\x10'::STRING,'1978-01-28':::DATE,NULL,','::\"char\",NULL,(-11):::INT8),('01020000A0E6100000020000006196ED5C911358C0109163650F8A3F40E0424C954E09F7C1F4D6D41A978959C03648187202BE42406E15181A2D3DE7C1':::GEOGRAPHY,NULL,'1982-02-14':::DATE,0.09532535507846451:::FLOAT8,'``'::\"char\",3.170364639645554522:::DECIMAL,(-113):::INT8)"
  ],
  "usesDeclarativeSchemaChanger": true
 }
}
    schema_change_external_test.go:131: 
        	Error Trace:	pkg/ccl/testccl/workload/schemachange/schema_change_external_test.go:131
        	Error:      	Received unexpected error:
        	            	***UNEXPECTED ERROR; Received an unexpected execution error.: ERROR: duplicate key value violates unique constraint "index_w0_68_w1_116" (SQLSTATE 23505)
        	            	(1) forced error mark
        	            	  | "fatal error when running txn"
        	            	  | github.com/cockroachdb/errors/withstack/*withstack.withStack::
        	            	Wraps: (2)
        	            	Wraps: (3) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*opStmt).executeStmt
        	            	  | 	pkg/workload/schemachange/operation_generator.go:3253
        	            	  | github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*schemaChangeWorker).runInTxn
        	            	  | 	pkg/workload/schemachange/schemachange.go:522
        	            	  | github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*schemaChangeWorker).run
        	            	  | 	pkg/workload/schemachange/schemachange.go:610
        	            	  | pkg/ccl/testccl/workload/schemachange/schemachange_test_test.TestWorkload.TestWorkload.func4.func6
        	            	  | 	pkg/ccl/testccl/workload/schemachange/schema_change_external_test.go:120
        	            	  | golang.org/x/sync/errgroup.(*Group).Go.func1
        	            	  | 	external/org_golang_x_sync/errgroup/errgroup.go:78
        	            	  | runtime.goexit
        	            	  | 	src/runtime/asm_amd64.s:1695
        	            	Wraps: (4) ***UNEXPECTED ERROR; Received an unexpected execution error.
        	            	Wraps: (5) ERROR: duplicate key value violates unique constraint "index_w0_68_w1_116" (SQLSTATE 23505)
        	            	Error types: (1) *markers.withMark (2) *schemachange.ErrorState (3) *withstack.withStack (4) *errutil.withPrefix (5) *pgconn.PgError
        	Test:       	TestWorkload
Schema Workload Stats
Total Schema Statements Executed = 124
Total Schema Statements Succeeded = 124
Total Schema Statement Expected Failures = 1
Total Transactions Committed = 87
Total Transactions Rolled Back = 72
Total Transactions Executed = 159
    schema_change_external_test.go:96: backup, tracing data, and system table dumps in /var/lib/engflow/worker/work/2/exec/bazel-out/k8-fastbuild/testlogs/pkg/ccl/testccl/workload/schemachange/schemachange_test/run_7_of_10/test.outputs/logTestWorkload3501019587
    panic.go:626: -- test log scope end --
test logs left over in: outputs.zip/logTestWorkload3501019587
--- FAIL: TestWorkload (12.82s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=7</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #142120 pkg/ccl/testccl/workload/schemachange/schemachange_test: TestWorkload failed [C-test-failure O-robot P-2 T-sql-foundations branch-release-25.1.1-rc]
- #141696 pkg/ccl/testccl/workload/schemachange/schemachange_test: TestWorkload failed [C-test-failure O-robot P-2 T-sql-foundations branch-release-23.2.21-rc]
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestWorkload.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-48368

--- a/pkg/workload/schemachange/error_screening.go
+++ b/pkg/workload/schemachange/error_screening.go
@@ -1774,3 +1774,32 @@ WITH tab_json AS (
 
 	return constraints, nil
 }
+
+// tableHasUniqueConstraintMutation determines if a table has any unique constraint
+// mutation ongoing. This means either being added or dropped.
+func (og *operationGenerator) tableHasUniqueConstraintMutation(
+	ctx context.Context, tx pgx.Tx, tableName *tree.TableName,
+) (bool, error) {
+	return og.scanBool(ctx, tx, `
+		WITH table_desc AS (
+			SELECT crdb_internal.pb_to_json(
+				'desc',
+				descriptor,
+				false
+			)->'table' as d
+			FROM system.descriptor
+			WHERE id = $1::REGCLASS
+		)
+		SELECT EXISTS (
+			SELECT * FROM (
+			SELECT jsonb_array_elements(
+				CASE WHEN d->'mutations' IS NULL
+				THEN '[]'::JSONB
+				ELSE d->'mutations'
+				END
+			) as m
+			FROM table_desc)
+			WHERE (m->>'direction')::STRING IN ('ADD', 'DROP')
+			AND (m->'index'->>'unique')::BOOL IS TRUE
+		);`, tableName)
+}

--- a/pkg/workload/schemachange/operation_generator.go
+++ b/pkg/workload/schemachange/operation_generator.go
@@ -3021,6 +3021,7 @@ func (og *operationGenerator) insertRow(ctx context.Context, tx pgx.Tx) (stmt *o
 	// Only evaluate these if we know that the inserted values are sane, since
 	// we will need to evaluate generated expressions below.
 	hasUniqueConstraints := false
+	hasUniqueConstraintsMutations := false
 	fkViolation := false
 	if !anyInvalidInserts {
 		// Verify if the new row may violate unique constraints by checking the
@@ -3030,6 +3031,12 @@ func (og *operationGenerator) insertRow(ctx context.Context, tx pgx.Tx) (stmt *o
 			return nil, err
 		}
 		hasUniqueConstraints = len(constraints) > 0
+		// If we have any unique constraint mutations pending then its always possible,
+		// for us to still violate them.
+		hasUniqueConstraintsMutations, err = og.tableHasUniqueConstraintMutation(ctx, tx, tableName)
+		if err != nil {
+			return nil, err
+		}
 		// Verify if the new row will violate fk constraints by checking the constraints and rows
 		// in the database.
 		fkViolation, err = og.violatesFkConstraints(ctx, tx, tableName, nonGeneratedColNames, rows)
@@ -3040,7 +3047,7 @@ func (og *operationGenerator) insertRow(ctx context.Context, tx pgx.Tx) (stmt *o
 	}
 
 	stmt.potentialExecErrors.addAll(codesWithConditions{
-		{code: pgcode.UniqueViolation, condition: hasUniqueConstraints},
+		{code: pgcode.UniqueViolation, condition: hasUniqueConstraints || hasUniqueConstraintsMutations},
 		{code: pgcode.ForeignKeyViolation, condition: fkViolation},
 		{code: pgcode.NotNullViolation, condition: true},
 		{code: pgcode.CheckViolation, condition: true},

